source/lib.ts

```diff
@@ -76,87 +76,96 @@ const promisifyStore = (passedStore: LegacyStore | Store): Store => {
 	return new PromisifiedStore()
 }
 
-type OptionsWithLegacyStore = Omit<Partial<Options>, 'store'> & {
-	store?: Store | LegacyStore
+// Configuration represents what actually ends up configuring the rate limiter. This is
+// derived from the users options, but with some legacy fields left ignored.
+type Configuration = Omit<
+	Options,
+	'headers' | 'draft_polli_ratelimit_headers' | 'store'
+> & {
+	store: Store
 }
 
 /**
  * Type-checks and adds the defaults for options the user has not specified.
  *
  * @param options {Options} - The options the user specifies.
  *
- * @returns {Options} - A complete configuration object.
+ * @returns {Configuration} - A complete configuration object.
  */
-const parseOptions = (passedOptions: OptionsWithLegacyStore): Options => {
+const parseOptions = (passedOptions: Partial<Options>): Configuration => {
 	// Passing undefined should be equivalent to not passing an option at all, so we'll
 	// omit all fields where their value is undefined.
-	const omittedOptions: OptionsWithLegacyStore =
+	const notUndefinedOptions: Partial<Options> =
 		omitUndefinedOptions(passedOptions)
 
 	// See ./types.ts#Options for a detailed description of the options and their
 	// defaults.
-	const options = {
-		windowMs: 60 * 1000,
-		store: new MemoryStore(),
-		max: 5,
-		message: 'Too many requests, please try again later.',
-		statusCode: 429,
-		legacyHeaders: passedOptions.headers ?? true,
-		standardHeaders: passedOptions.draft_polli_ratelimit_headers ?? false,
-		requestPropertyName: 'rateLimit',
-		skipFailedRequests: false,
-		skipSuccessfulRequests: false,
-		requestWasSuccessful: (_request: Request, response: Response): boolean =>
-			response.statusCode < 400,
-		skip: (_request: Request, _response: Response): boolean => false,
-		keyGenerator: (request: Request, _response: Response): string => {
-			if (!request.ip) {
-				console.error(
-					'WARN | `express-rate-limit` | `request.ip` is undefined. You can avoid this by providing a custom `keyGenerator` function, but it may be indicative of a larger issue.',
-				)
-			}
+	const config: Omit<Configuration, 'store'> & { store: Store | LegacyStore } =
+		{
+			windowMs: 60 * 1000,
+			// It's possible that this field might have a LegacyConfig if overrode by passedOptions.
+			// However, LegacyStores are converted to Stores later.
+			store: new MemoryStore(),
```

Comment:
I don't like the `Omit` stuff we are doing [here](https://github.com/nfriedly/express-rate-limit/pull/294/files/6d2632fbc6f5f5d236e2725b4002be38d4649ce0..7eb1e6b39a2ff1dfdb92d65729bff9168f2b41aa#diff-5c2516e9822679491492c832b9ace79b6945bdd891586e07b0232ebef1c7e0b1R103) - we could remove the need for that by changing this line to:

```suggestion
			store: promisifyStore(notUndefinedOptions.store ?? new MemoryStore()),
```

And removing the `promisifyStore` call [here](https://github.com/nfriedly/express-rate-limit/pull/294/files/6d2632fbc6f5f5d236e2725b4002be38d4649ce0..7eb1e6b39a2ff1dfdb92d65729bff9168f2b41aa#diff-5c2516e9822679491492c832b9ace79b6945bdd891586e07b0232ebef1c7e0b1R165).